scarlet/psf.py

```diff
@@ -1,268 +1,243 @@
+
+        super().__init__(*parameters)
+
+        self.integrate = integrate
+
+        if boxsize is None:
+            boxsize = 15
+        if boxsize % 2 == 0:
+            boxsize += 1
+
+        # length of 0 parameter gives number of channels
+        p0 = self.get_parameter(0, *parameters)
+        shape = (len(p0), boxsize, boxsize)
+        origin = (0, -(boxsize // 2), -(boxsize // 2))
+        self.bbox = Box(shape, origin=origin)
 
-def gaussian(y, x, sigma=1, integrate=True, bbox=None):
+        self._Y = np.arange(self.bbox.shape[-2]) + self.bbox.origin[-2]
+        self._X = np.arange(self.bbox.shape[-1]) + self.bbox.origin[-1]
+
+        # same across all bands
+        self.is_same = np.all(p0 == p0[0])
+        self._d = self.bbox.D - 2
+
+    def expand_dims(self, model):
+        return np.expand_dims(model, axis=np.arange(self._d))
+
+
+class GaussianPSF(FunctionPSF):
     """Circular Gaussian Function
 
     Parameters
     ----------
-    y: float
-        Vertical coordinate of the center
-    x: float
-        Horizontal coordinate of the center
-    sigma: float
-        Standard deviation of the gaussian
+    sigma: float, array, or `~scarlet.Parameter`
+        Standard deviation of the Gaussian in frame pixels
+        If the width is to be optimized, provide a full defined `Parameter`.
     integrate: bool
         Whether pixel integration is performed
-    bbox: Box
-        Bounding box over which to evaluate the function
-
-    Returns
-    -------
-    result: array
-        A 2D circular gaussian sampled at the coordinates `(y_i, x_j)`
-        for all i and j in `shape`.
+    boxsize: Box
+        Size of bounding box over which to evaluate the function, in frame pixels
     """
-    Y = np.arange(bbox.shape[1]) + bbox.origin[1]
-    X = np.arange(bbox.shape[2]) + bbox.origin[2]
 
-    def f(X):
-        if not integrate:
+    def __init__(self, sigma, integrate=True, boxsize=None):
+
+        sigma = self.prepare_param(sigma, "sigma")
+
+        if boxsize is None:
+            boxsize = int(np.ceil(10 * np.max(sigma)))
+
+        super().__init__(sigma, integrate=integrate, boxsize=boxsize)
+
+    def get_model(self, *parameters, offset=None):
+
+        sigma = self.get_parameter(0, *parameters)
+
+        if offset is None:
+            offset = (0, 0)
+
+        if self.is_same:
+            s = sigma[0]
+            psfs = self.expand_dims(
+                self._f(self._Y - offset[0], s)[:, None]
+                * self._f(self._X - offset[1], s)[None, :]
+            )
+        else:
+            psfs = np.stack(
+                (
+                    self._f(self._Y - offset[0], s)[:, None]
+                    * self._f(self._X - offset[1], s)[None, :]
+                    for s in sigma
+                ),
+                axis=0,
+            )
+        # use image integration instead of analytic for consistency with other PSFs
+        return normalize(psfs)
+
+    def _f(self, X, sigma):
```

Comment:
Not sure it needs to be a class method.

Reply:
sure. `self.integrate`. Also, where else would you put it?